packages/autocomplete-plugin-redirect-url/src/createRedirectUrlPlugin.ts

```diff
@@ -0,0 +1,185 @@
+
+import { defaultTemplates } from './templates';
+import {
+  RedirectUrlItem,
+  RedirectUrlPlugin as RedirectUrlPluginData,
+} from './types';
+
+export type OnRedirectOptions<TItem extends RedirectUrlItem> = {
+  navigator: InternalAutocompleteOptions<TItem>['navigator'];
+  state: AutocompleteState<TItem>;
+};
+
+type TransformResponseParams<TItem> =
+  | SearchResponse<TItem>
+  | SearchForFacetValuesResponse;
+
+export type CreateRedirectUrlPluginParams<TItem extends BaseItem> = {
+  transformResponse?(
+    response: TransformResponseParams<TItem>
+  ): string | undefined;
+  onRedirect?(
+    redirects: RedirectUrlItem[],
+    options: OnRedirectOptions<RedirectUrlItem>
+  ): void;
+  templates?: SourceTemplates<RedirectUrlItem>;
+};
+
+function defaultTransformResponse<THit>(
+  response: TransformResponseParams<THit>
+): string | undefined {
+  return (response as Record<string, any>).renderingContent?.redirect?.url;
+}
+
+function defaultOnRedirect(
+  redirects: RedirectUrlItem[],
+  { navigator, state }: OnRedirectOptions<RedirectUrlItem>
+) {
+  const itemUrl = redirects[0]?.urls?.[0];
+  if (itemUrl) {
+    navigator.navigate({ itemUrl, item: redirects[0], state });
+  }
+}
+
+function getOptions<TItem extends BaseItem>(
+  options: CreateRedirectUrlPluginParams<TItem>
+) {
+  return {
+    transformResponse: defaultTransformResponse,
+    templates: defaultTemplates,
+    onRedirect: defaultOnRedirect,
+    ...options,
+  };
+}
+
+function getRedirectData({ state }) {
+  const redirectUrlPlugin = state.context
+    .redirectUrlPlugin as RedirectUrlPluginData;
+  return (redirectUrlPlugin?.data ?? []) as RedirectUrlItem[];
+}
+
+export function createRedirectUrlPlugin<TItem extends BaseItem>(
+  options: CreateRedirectUrlPluginParams<TItem>
+): AutocompletePlugin<RedirectUrlItem, undefined> {
+  const { transformResponse, templates, onRedirect } = getOptions(options);
+
+  function createRedirects({ results, source, state }): RedirectUrlItem[] {
+    const redirect: RedirectUrlItem = {
+      sourceId: source.sourceId,
+      urls: results
+        .map((result) => transformResponse(result))
+        .filter((url) => url !== undefined),
+    };
+
+    const redirects: RedirectUrlItem[] =
+      state.context.redirectUrlPlugin?.data ?? [];
+    const existingRedirectIndex = redirects.findIndex(
+      (r) => r.sourceId === source.sourceId
+    );
+
+    if (existingRedirectIndex !== -1) {
+      if (redirect.urls.length === 0) {
+        redirects.splice(existingRedirectIndex, 1);
+      } else {
+        redirects[existingRedirectIndex] = redirect;
+      }
+    } else if (redirect.urls.length > 0) {
+      redirects.push(redirect);
+    }
+
+    return redirects;
+  }
+
+  let navigator: InternalAutocompleteOptions<RedirectUrlItem>['navigator'];
+
+  return {
+    name: 'aa.redirectUrlPlugin',
+    subscribe({ onResolve, setContext, props }) {
+      navigator = props.navigator;
```

Comment:
if accessing the navigator was possible in reshape / onSelect / onSubmit, we wouldn't need to expose props here. Is that a change we should make?

Reply:
I agree. This would prevent relying on an outer-scope variable as well, which feels a tad hacky to me.

(when this exits draft) Do any other reviewers agree or disagree with this though?

Reply:
I agree it feels like a better contract.